src/units/abilities.cpp

```diff
@@ -1115,60 +1130,91 @@ unit_ability_list attack_type::get_weapon_ability(const std::string& ability) co
-			if(!((*i->ability_cfg)["overwrite_specials"].to_bool() && special_active_impl(other_attack_, shared_from_this(), *i->ability_cfg, AFFECT_OTHER, ability, true, "filter_student"))) {
+			if(!special_active_impl(other_attack_, shared_from_this(), *i->ability_cfg, AFFECT_OTHER, ability, true, "filter_student")) {
 				i = abil_other_list.erase(i);
 			} else {
 				++i;
 			}
 		}
-		if(abil_other_list.empty()){
-			abil_other_list.append((*other_).get_abilities(ability, other_loc_));
-			for(unit_ability_list::iterator i = abil_other_list.begin(); i != abil_other_list.end();) {
-				if(!special_active_impl(other_attack_, shared_from_this(), *i->ability_cfg, AFFECT_OTHER, ability, true, "filter_student")) {
-					i = abil_other_list.erase(i);
-				} else {
-					++i;
-				}
-			}
-		}
 	}
+
 	abil_list.append(abil_other_list);
+	if(!abil_list.empty()){
+		abil_list = overwrite_special_checking(ability, abil_list, abil_list, "filter_student");
+	}
 	return abil_list;
 }
 
 unit_ability_list attack_type::get_specials_and_abilities(const std::string& special) const
 {
 	unit_ability_list abil_list = get_weapon_ability(special);
+	unit_ability_list spe_list = get_specials(special);
+	if(!spe_list.empty()){
+		spe_list = overwrite_special_checking(special, spe_list, abil_list);
+		if(special == "plague" && !spe_list.empty()){
+			return spe_list;
+		}
+		abil_list.append(spe_list);
+	}
+	return abil_list;
+}
+
+unit_ability_list attack_type::overwrite_special_checking(const std::string& ability, unit_ability_list temp_list, unit_ability_list abil_list, const std::string& filter_self) const
+{
+	bool overwrite_self = false;
+	bool overwrite_either = false;
+
 	for(unit_ability_list::iterator i = abil_list.begin(); i != abil_list.end();) {
-		if((*i->ability_cfg)["overwrite_specials"].to_bool()) {
-			return abil_list;
+		if((*i->ability_cfg)["overwrite_specials"] == "both_sides") {
+			overwrite_either = true;
+			break;
+		}
+		if((overwrite_special_affects(*i->ability_cfg) && special_active(*i->ability_cfg, AFFECT_SELF, ability, true, "filter_student")) || (overwrite_special_affects(*i->ability_cfg) && special_active_impl(other_attack_, shared_from_this(), *i->ability_cfg, AFFECT_OTHER, ability, true, "filter_student"))) {
+			overwrite_self = true;
 		}
 		++i;
 	}
-	abil_list.append(get_specials(special));
-	return abil_list;
+	if(!overwrite_either && !overwrite_self){
+		return temp_list;
+	} else if(overwrite_either){
+		for(unit_ability_list::iterator i = temp_list.begin(); i != temp_list.end();) {
+			bool overwrite = false;
+			if(filter_self == "filter_student"){
+				overwrite = overwrite_special_affects(*i->ability_cfg);
+			}
+			if((special_active(*i->ability_cfg, AFFECT_SELF, ability, true, filter_self) && !overwrite) || (special_active_impl(other_attack_, shared_from_this(), *i->ability_cfg, AFFECT_OTHER, ability, true, filter_self) && !overwrite)) {
```

Comment:
This can be shortened to `!overwrite && (special_active(*i->ability_cfg, AFFECT_SELF, ability, true, filter_self) || special_active_impl(other_attack_, shared_from_this(), *i->ability_cfg, AFFECT_OTHER, ability, true, filter_self))`.